src/cdomains/structDomain.ml

```diff
@@ -90,3 +105,496 @@ struct
+            aux (joined::rem_uniq) t
+        in
+        let res = aux [] (HS.elements s) in
+        if Messages.tracing then Messages.tracel "reduce-key" "Reduced - s:\n%a\nto:\n%a\n---------\n" HS.pretty s HS.pretty res;
+        (res, Some key)
+
+  let reduce_key (x: t): t = reduce_key_with_fct (SS.join) x
+
+  let replace (x: t) field value : t =
+    let (s, _) = x in
+    if Messages.tracing then Messages.tracel "keyedsets" "Replace - s:\n%a\nfield:%a\nvalue: %a\n---------\n" HS.pretty s Basetype.CilField.pretty field Val.pretty value ;
+    let replaced = HS.map (fun s -> SS.replace s field value) s in
+    let result_key =
+      match find_key_field x with
+      | None -> (replaced, None)
+      | Some key ->
+        if Basetype.CilField.equal key field
+        then (join_set replaced, Some key) (* Key is now the same in all variants *)
+        else (replaced, Some key)
+    in
+    result_key
+
+  let get (s, _: t) field =
+    HS.fold (fun ss acc -> Val.join acc (SS.get ss field)) s (Val.bot ())
+
+  let fold f (s, _: t) a = on_joint_ss (fun ss -> SS.fold f ss a) a s
+
+  let map f (x: t) =
+    let (s, k) = x in
+    if HS.is_bot s
+    then x
+    else
+      let s' = HS.singleton (on_joint_ss (SS.map f) (SS.top ()) s) in
+      reduce_key (s', k)
+
+  let equal (x,_) (y,_) = HS.equal x y
+  let compare (x,_) (y,_) = HS.compare x y
+
+  let take_some_key (k1: field option) (k2: field option) (s: set) =
+    match (k1, k2) with
+    | (Some(x), _) -> Some x
+    | (_, Some(y)) -> Some y
+    | _ -> find_key_field (s, None)
+
+  let meet_narrow_common x y f =
+    (*
+    Normally, this would be just: HS.meet x y |> filter_variants_with_bot
+    However, arrays set everything at start to bottom, and we would quickly have
+    some things initialized to bottom, and discard them.
+    That's why we take a more involved approach and only discard those variants
+    that were not overlapped before (non-bottom meet && none of them was bottom before)
+     *)
+    begin
+      let fields_overlap a b =
+        Val.is_bot_value a || Val.is_bot_value b || not (Val.is_bot_value (Val.meet a b))
+      in
+      let maps_overlap ssx ssy =
+        let fields = SS.keys ssx in
+        List.for_all (fun field -> fields_overlap (SS.get ssx field) (SS.get ssy field)) fields
+      in
+      let ((sx, kx), (sy, ky)) = (x, y) in
+      let x_list = HS.elements sx in
+      let y_list = HS.elements sy in
+      let s = List.map (fun xss -> List.map (fun yss -> (xss, yss)) y_list) x_list
+              |> List.flatten
+              |> List.filter (fun (ssx, ssy) -> maps_overlap ssx ssy)
+              |> List.map (fun (ssx, ssy) -> f ssx ssy)
+              |> HS.of_list
+      in reduce_key (s, take_some_key kx ky s)
+    end
+
+  let meet x y = meet_narrow_common x y SS.meet
+
+  let join (x, kx) (y, ky) = let s = HS.join x y in reduce_key (s, take_some_key kx ky s)
+  let leq (x, _) (y, _) = HS.leq x y
```

Comment:
This is suspicious. How can the second component not matter for `leq` if it does for `join`? This violates a lattice property.

Reply:
It is deterministically the same for the same `compinfo`. So there should never be any joins with different keys.

Reply:
The reason all of this crap is needed at all because all structs (regardless of their type) are represented as a map, where e.g. bot is `{}`